File: Polars.CSharp/IO/Parquet/Read.cs
#pragma warning disable CS1573
using Polars.NET.Core;
using Cs = Polars.CSharp.Polars.Selectors;

namespace Polars.CSharp;
public partial class DataFrame : IDisposable,IEnumerable<Series>,IPolarsDataFrame
{
    /// <summary>
    /// Read a DataFrame from a Parquet file or multiple files via glob patterns.
    /// <para>
    /// Note: This method internally uses LazyFrame.ScanParquet and collects the result. 
    /// For larger-than-memory datasets or better query optimization, consider using LazyFrame.ScanParquet directly.
    /// </para>
    /// </summary>
    /// <param name="path">Path to file or glob pattern (e.g. "data/*.parquet" or "s3://bucket/data.parquet").</param>
    /// <param name="columns">Columns to select. If null, select all columns. (Optimized via projection pushdown).</param>
    /// <param name="nRows">Limit number of rows to read (optimization).</param>
    /// <param name="parallel">Parallel strategy.</param>
    /// <param name="lowMemory">Reduce memory usage at the expense of performance.</param>
    /// <param name="useStatistics">Use parquet statistics to optimize the query plan.</param>
    /// <param name="glob">Expand glob patterns (default: true).</param>
    /// <param name="allowMissingColumns">Allow missing columns when reading multiple files.</param>
    /// <param name="rechunk">Rechunk the memory to contiguous chunks when reading. (default: false)</param>
    /// <param name="cache">Cache the result after reading. (default: true)</param>
    /// <param name="rowIndexName">If provided, adds a column with the row number.</param>
    /// <param name="rowIndexOffset">Offset for the row index.</param>
    /// <param name="includePathColumn">If provided, adds a column with the source file path.</param>
    /// <param name="schema">Manually specify the schema of the file(s).</param>
    /// <param name="hivePartitionSchema">Manually specify the schema for Hive partitioning columns.</param>
    /// <param name="tryParseHiveDates">Whether to try parsing dates in Hive partitioning paths (default: true).</param>
    /// <param name="cloudOptions">Options for cloud storage (AWS S3, Azure Blob, GCS, etc.).</param>
    public static DataFrame ReadParquet(
        string path,
        string[]? columns = null,
        ulong? nRows = null,
        ParallelStrategy parallel = ParallelStrategy.Auto,
        bool lowMemory = false,
        bool useStatistics = true,
        bool glob = true,
        bool allowMissingColumns = false,
        bool rechunk = false, 
        bool cache = true,    
        string? rowIndexName = null,
        uint rowIndexOffset = 0,
        string? includePathColumn = null,
        PolarsSchema? schema = null,
        bool hivePartitioning = false,
        PolarsSchema? hivePartitionSchema = null,
        bool tryParseHiveDates = false,
        CloudOptions? cloudOptions = null)
    {
        var lf = LazyFrame.ScanParquet(
            path,
            nRows,
            parallel,
            lowMemory,
            useStatistics,
            glob,
            allowMissingColumns,
            rechunk,
            cache,
            rowIndexName,
            rowIndexOffset,
            includePathColumn,
            schema,
            hivePartitioning,
            hivePartitionSchema,
            tryParseHiveDates,
            cloudOptions
        );

        if (columns != null && columns.Length > 0)
        {
            var colsToSelect = new List<string>(columns);
            
            if (!string.IsNullOrEmpty(rowIndexName) && !colsToSelect.Contains(rowIndexName))
            {
                colsToSelect.Add(rowIndexName);
            }

            if (!string.IsNullOrEmpty(includePathColumn) && !colsToSelect.Contains(includePathColumn))
            {
                colsToSelect.Add(includePathColumn);
            }

            lf = lf.Select(Cs.ByName([.. colsToSelect]));
        }

        return lf.Collect();
    }

    /// <summary>
    /// Read Parquet from an in-memory byte array.
    /// <para>
    /// Note: This method internally uses LazyFrame.ScanParquet and collects the result.
    /// </para>
    /// </summary>
    /// <param name="buffer">The byte array containing the parquet file.</param>
    /// <param name="columns">Columns to select. If null, select all columns.</param>
    /// <inheritdoc cref="ReadParquet(string, string[], ulong?, ParallelStrategy, bool, bool, bool, bool, bool, bool, string?, uint, string?, PolarsSchema?,bool, PolarsSchema?, bool, CloudOptions?)"/>
    public static DataFrame ReadParquet(
        byte[] buffer,
        string[]? columns = null,
        ulong? nRows = null,
        ParallelStrategy parallel = ParallelStrategy.Auto,
        bool lowMemory = false,
        bool useStatistics = true,
        bool allowMissingColumns = false,
        bool rechunk = false, 
        bool cache = true,    
        string? rowIndexName = null,
        uint rowIndexOffset = 0,
        string? includePathColumn = null,
        PolarsSchema? schema = null,
        bool hivePartitioning = false,
        PolarsSchema? hivePartitionSchema = null,
        bool tryParseHiveDates = false)
    {
        var lf = LazyFrame.ScanParquet(
            buffer,
            nRows,
            parallel,
            lowMemory,
            useStatistics,
            allowMissingColumns,
            rechunk,
            cache,
            rowIndexName,
            rowIndexOffset,
            includePathColumn,
            schema,
            hivePartitioning,
            hivePartitionSchema,
            tryParseHiveDates
        );

        if (columns != null && columns.Length > 0)
        {
            var colsToSelect = new List<string>(columns);
            
            if (!string.IsNullOrEmpty(rowIndexName) && !colsToSelect.Contains(rowIndexName))
            {
                colsToSelect.Add(rowIndexName);
            }

            if (!string.IsNullOrEmpty(includePathColumn) && !colsToSelect.Contains(includePathColumn))
            {
                colsToSelect.Add(includePathColumn);
            }

            lf = lf.Select(Cs.ByName([.. colsToSelect]));
        }

        return lf.Collect();
    }

    /// <summary>
    /// Read Parquet from a Stream.
    /// <para>
    /// Note: This method copies the stream to memory and then uses the Lazy execution engine.
    /// </para>
    /// </summary>
    /// <param name="stream">The input stream containing the parquet file.</param>
    /// <param name="columns">Columns to select. If null, select all columns.</param>
    /// <inheritdoc cref="ReadParquet(string, string[], ulong?, ParallelStrategy, bool, bool, bool, bool, bool, bool, string?, uint, string?, PolarsSchema?,bool, PolarsSchema?, bool, CloudOptions?)"/>
    public static DataFrame ReadParquet(
        Stream stream,
        string[]? columns = null,
        ulong? nRows = null,
        ParallelStrategy parallel = ParallelStrategy.Auto,
        bool lowMemory = false,
        bool useStatistics = true,
        bool allowMissingColumns = false,
        bool rechunk = false, 
        bool cache = true,    
        string? rowIndexName = null,
        uint rowIndexOffset = 0,
        string? includePathColumn = null,
        PolarsSchema? schema = null,
        bool hivePartitioning = false,
        PolarsSchema? hivePartitionSchema = null,
        bool tryParseHiveDates = false)
    {
        using var ms = new MemoryStream();
        stream.CopyTo(ms);
        
        return ReadParquet(
            ms.ToArray(),
            columns,
            nRows,
            parallel,
            lowMemory,
            useStatistics,
            allowMissingColumns,
            rechunk,
            cache,
            rowIndexName,
            rowIndexOffset,
            includePathColumn,
            schema,
            hivePartitioning,
            hivePartitionSchema,
            tryParseHiveDates
        );
    }
    /// <summary>
    /// Read a Parquet file asynchronously.
    /// </summary>
    public static async Task<DataFrame> ReadParquetAsync(
        string path,
        string[]? columns = null,
        ulong? nRows = null,
        ParallelStrategy parallel = ParallelStrategy.Auto,
        bool lowMemory = false,
        bool useStatistics = true,
        bool glob = true,
        bool allowMissingColumns = false,
        bool rechunk = false, 
        bool cache = true,    
        string? rowIndexName = null,
        uint rowIndexOffset = 0,
        string? includePathColumn = null,
        PolarsSchema? schema = null,
        bool hivePartitioning = false,
        PolarsSchema? hivePartitionSchema = null,
        bool tryParseHiveDates = false,
        CloudOptions? cloudOptions = null)
    {
        var lf = LazyFrame.ScanParquet(
            path,
            nRows,
            parallel,
            lowMemory,
            useStatistics,
            glob,
            allowMissingColumns,
            rechunk,
            cache,
            rowIndexName,
            rowIndexOffset,
            includePathColumn,
            schema,
            hivePartitioning,
            hivePartitionSchema,
            tryParseHiveDates,
            cloudOptions
        );

        if (columns != null && columns.Length > 0)
        {
            var colsToSelect = new List<string>(columns);
            
            if (!string.IsNullOrEmpty(rowIndexName) && !colsToSelect.Contains(rowIndexName))
            {
                colsToSelect.Add(rowIndexName);
            }

            if (!string.IsNullOrEmpty(includePathColumn) && !colsToSelect.Contains(includePathColumn))
            {
                colsToSelect.Add(includePathColumn);
            }

            lf = lf.Select(Cs.ByName([.. colsToSelect]));
        }

        return await lf.CollectAsync(useStreaming:true);
    }
}

File: Polars.FSharp/IO/Parquet/Read.fs
namespace Polars.FSharp

open Polars.NET.Core

[<AutoOpen>]
module ParquetRead =
    open System.IO
    type LazyFrame with
        /// <summary> Scan a parquet file into a LazyFrame. </summary>
        /// <summary>
        /// Lazily read from a parquet file or a common cloud store (S3, GCS, Azure, etc.).
        /// </summary>
        /// <param name="path">Path to file or cloud location.</param>
        /// <param name="nRows">Stop reading after n rows.</param>
        /// <param name="parallel">Parallel strategy (Auto, Columns, RowGroups, None).</param>
        /// <param name="lowMemory">Reduce memory pressure at the expense of performance.</param>
        /// <param name="useStatistics">Use parquet statistics to prune row groups.</param>
        /// <param name="glob">Expand path using globbing rules.</param>
        /// <param name="allowMissingColumns">If true, do not fail if columns are missing.</param>
        /// <param name="rechunk">Rechunk the memory to contiguous chunks when reading. (default: false)</param>
        /// <param name="cache">Cache the result after reading. (default: true)</param>
        /// <param name="rowIndexName">If provided, add a row index column with this name.</param>
        /// <param name="rowIndexOffset">Start index for the row index column.</param>
        /// <param name="includePathColumn">If provided, add a column with the path of the file.</param>
        /// <param name="schema">Overwrite the schema of the dataset.</param>
        /// <param name="hivePartitionSchema">The schema of the hive partitions.</param>
        /// <param name="tryParseHiveDates">Attempt to parse hive values as Date/Datetime.</param>
        /// <param name="cloudOptions">Options for cloud storage (AWS S3, Azure Blob, GCS, etc.).</param>
        static member ScanParquet
            (
                path: string,
                ?nRows: uint64,
                ?parallelStrategy: ParallelStrategy,
                ?lowMemory: bool,
                ?useStatistics: bool,
                ?glob: bool,
                ?allowMissingColumns: bool,
                ?rechunk:bool,
                ?cache:bool,
                ?rowIndexName: string,
                ?rowIndexOffset: uint32,
                ?includePathColumn: string,
                ?schema: PolarsSchema,
                ?hivePartitioning: bool,
                ?hivePartitionSchema: PolarsSchema,
                ?tryParseHiveDates: bool,
                ?cloudOptions: CloudOptions
            ) =
            // Defaults
            let pParallel = defaultArg parallelStrategy ParallelStrategy.Auto
            let pLowMem = defaultArg lowMemory false
            let pStats = defaultArg useStatistics true
            let pGlob = defaultArg glob true
            let pAllowMissing = defaultArg allowMissingColumns false
            let pRowIndexOffset = defaultArg rowIndexOffset 0u
            let pTryHive = defaultArg tryParseHiveDates false
            let pRechunk = defaultArg rechunk false
            let pCache = defaultArg cache false
            let pHivePartitioning = defaultArg hivePartitioning false

            // F# Types -> C# Interop Types
            
            // nRows: uint64 option -> ulong? (Nullable<ulong>)
            let pNRows = 
                nRows 
                |> Option.toNullable

            // Schema: Schema Object -> SchemaHandle (Raw Pointer holder)
            let hSchema = 
                schema 
                |> Option.map (fun s -> s.Handle) 
                |> Option.toObj

            let hHiveSchema = 
                hivePartitionSchema 
                |> Option.map (fun s -> s.Handle) 
                |> Option.toObj

            // Cloud Options Unwrapping Logic
            let cProv, cRet, cToMs, cInitMs, cMaxMs, cCache, cKeys, cVals = 
                CloudOptions.ParseCloudOptions cloudOptions

            // Call Wrapper
            let handle = PolarsWrapper.ScanParquet(
                path,
                pNRows,
                pParallel.ToNative(),
                pLowMem,
                pStats,
                pGlob,
                pAllowMissing,
                pRechunk,
                pCache,
                Option.toObj rowIndexName,
                pRowIndexOffset,
                Option.toObj includePathColumn,
                hSchema,
                pHivePartitioning,
                hHiveSchema,
                pTryHive,
                cProv,
                cRet,
                cToMs,
                cInitMs,
                cMaxMs,
                cCache,
                cKeys,
                cVals
            )
            
            new LazyFrame(handle)
        /// <summary>
        /// [Memory] Lazily read parquet from a byte array (in-memory buffer).
        /// </summary>
        /// <param name="buffer">The byte array containing parquet data.</param>
        /// <param name="nRows">Stop reading after n rows.</param>
        /// <param name="parallel">Parallel strategy (Auto, Columns, RowGroups, None).</param>
        /// <param name="lowMemory">Reduce memory pressure at the expense of performance.</param>
        /// <param name="useStatistics">Use parquet statistics to prune row groups.</param>
        /// <param name="glob">Globbing patterns (usually irrelevant for memory scan,always false).</param>
        /// <param name="allowMissingColumns">If true, do not fail if columns are missing.</param>
        /// <param name="rechunk">Rechunk the memory to contiguous chunks when reading. (default: false)</param>
        /// <param name="cache">Cache the result after reading. (default: true)</param>
        /// <param name="rowIndexName">If provided, add a row index column with this name.</param>
        /// <param name="rowIndexOffset">Start index for the row index column.</param>
        /// <param name="includePathColumn">If provided, add a column with the path (usually irrelevant for memory).</param>
        /// <param name="schema">Overwrite the schema of the dataset.</param>
        /// <param name="hivePartitionSchema">The schema of the hive partitions.</param>
        /// <param name="tryParseHiveDates">Attempt to parse hive values as Date/Datetime.</param>
        static member ScanParquet
            (
                buffer: byte[],
                ?nRows: uint64,
                ?parallelStrategy: ParallelStrategy,
                ?lowMemory: bool,
                ?useStatistics: bool,
                ?glob: bool,
                ?allowMissingColumns: bool,
                ?rechunk:bool,
                ?cache:bool,
                ?rowIndexName: string,
                ?rowIndexOffset: uint32,
                ?includePathColumn: string,
                ?schema: PolarsSchema,       
                ?hivePartitioning: bool,
                ?hivePartitionSchema: PolarsSchema,   
                ?tryParseHiveDates: bool
            ) =
            let pParallel = defaultArg parallelStrategy ParallelStrategy.Auto
            let pLowMem = defaultArg lowMemory false
            let pStats = defaultArg useStatistics true
            let _pGlob = defaultArg glob true
            let pAllowMissing = defaultArg allowMissingColumns false
            let pRowIndexOffset = defaultArg rowIndexOffset 0u
            let pTryHive = defaultArg tryParseHiveDates false
            let pRechunk = defaultArg rechunk false
            let pCache = defaultArg cache false
            let pHivePartitioning = defaultArg hivePartitioning false

            // 2. Type Conversions
            let pNRows = 
                nRows 
                |> Option.toNullable

            // Extract Handle from PolarsSchema wrapper
            let hSchema = 
                schema 
                |> Option.map (fun s -> s.Handle) 
                |> Option.toObj

            let hHiveSchema = 
                hivePartitionSchema 
                |> Option.map (fun s -> s.Handle) 
                |> Option.toObj

            // 3. Call C# Wrapper (Memory Overload)
            let handle = PolarsWrapper.ScanParquet(
                buffer,
                pNRows,
                pParallel.ToNative(),
                pLowMem,
                pStats,
                false,
                pAllowMissing,
                pRechunk,
                pCache,
                Option.toObj rowIndexName,
                pRowIndexOffset,
                Option.toObj includePathColumn,
                hSchema,
                pHivePartitioning,
                hHiveSchema,
                pTryHive
            )

            new LazyFrame(handle)
    type DataFrame with
           /// <summary>
        /// Read a parquet file into a DataFrame (Eager).
        /// <para>
        /// Note: This method internally uses LazyFrame.ScanParquet and collects the result.
        /// </para>
        /// </summary>
        static member ReadParquet(
            path: string,
            ?columns: seq<string>,
            ?nRows: uint64,
            ?parallelStrategy: ParallelStrategy,
            ?lowMemory: bool,
            ?useStatistics: bool,
            ?glob: bool,
            ?allowMissingColumns: bool,
            ?rechunk: bool,
            ?cache: bool,
            ?rowIndexName: string,
            ?rowIndexOffset: uint32,
            ?includePathColumn: string,
            ?schema: PolarsSchema,
            ?hivePartitioning: bool,
            ?hivePartitionSchema: PolarsSchema,
            ?tryParseHiveDates: bool,
            ?cloudOptions: CloudOptions
        ) : DataFrame =
            
            let mutable (lf: LazyFrame) = LazyFrame.ScanParquet(
                path,
                ?nRows = nRows,
                ?parallelStrategy = parallelStrategy,
                ?lowMemory = lowMemory,
                ?useStatistics = useStatistics,
                ?glob = glob,
                ?allowMissingColumns = allowMissingColumns,
                ?rechunk = rechunk,
                ?cache = cache,
                ?rowIndexName = rowIndexName,
                ?rowIndexOffset = rowIndexOffset,
                ?includePathColumn = includePathColumn,
                ?schema = schema,
                ?hivePartitioning = hivePartitioning,
                ?hivePartitionSchema = hivePartitionSchema,
                ?tryParseHiveDates = tryParseHiveDates,
                ?cloudOptions = cloudOptions
            )

            match columns with
            | Some cols -> 
                let mutable colList = cols |> Seq.toList
                if not (List.isEmpty colList) then
                    match rowIndexName with
                    | Some rName when not (List.contains rName colList) -> colList <- colList @ [rName]
                    | _ -> ()

                    match includePathColumn with
                    | Some pName when not (List.contains pName colList) -> colList <- colList @ [pName]
                    | _ -> ()

                    use sel = new Selector(PolarsWrapper.SelectorCols (colList |> List.toArray))
                    lf <- lf.Select [sel :> IColumnExpr]
            | None -> ()

            lf.Collect()

        // ---------------------------------------------------------
        // Read Parquet (Memory / Bytes)
        // ---------------------------------------------------------

        /// <summary>
        /// Read Parquet from an in-memory byte array.
        /// </summary>
        static member ReadParquet(
            buffer: byte[],
            ?columns: seq<string>,
            ?nRows: uint64,
            ?parallelStrategy: ParallelStrategy,
            ?lowMemory: bool,
            ?useStatistics: bool,
            ?allowMissingColumns: bool,
            ?rechunk: bool,
            ?cache: bool,
            ?rowIndexName: string,
            ?rowIndexOffset: uint32,
            ?includePathColumn: string,
            ?schema: PolarsSchema,
            ?hivePartitioning:bool,
            ?hivePartitionSchema: PolarsSchema,
            ?tryParseHiveDates: bool
        ) : DataFrame =
            
            let mutable (lf: LazyFrame) = LazyFrame.ScanParquet(
                buffer,
                ?nRows = nRows,
                ?parallelStrategy = parallelStrategy,
                ?lowMemory = lowMemory,
                ?useStatistics = useStatistics,
                ?allowMissingColumns = allowMissingColumns,
                ?rechunk = rechunk,
                ?cache = cache,
                ?rowIndexName = rowIndexName,
                ?rowIndexOffset = rowIndexOffset,
                ?includePathColumn = includePathColumn,
                ?schema = schema,
                ?hivePartitioning = hivePartitioning,
                ?hivePartitionSchema = hivePartitionSchema,
                ?tryParseHiveDates = tryParseHiveDates
            )

            match columns with
            | Some cols -> 
                let mutable colList = cols |> Seq.toList
                if not (List.isEmpty colList) then
                    match rowIndexName with
                    | Some rName when not (List.contains rName colList) -> colList <- colList @ [rName]
                    | _ -> ()

                    match includePathColumn with
                    | Some pName when not (List.contains pName colList) -> colList <- colList @ [pName]
                    | _ -> ()

                    use sel = new Selector(PolarsWrapper.SelectorCols(colList |> List.toArray))
                    lf <- lf.Select [sel :> IColumnExpr]
            | None -> ()

            lf.Collect()

        // ---------------------------------------------------------
        // Read Parquet (Stream)
        // ---------------------------------------------------------

        /// <summary>
        /// Read Parquet from a Stream.
        /// </summary>
        static member ReadParquet(
            stream: System.IO.Stream,
            ?columns: seq<string>,
            ?nRows: uint64,
            ?parallelStrategy: ParallelStrategy,
            ?lowMemory: bool,
            ?useStatistics: bool,
            ?allowMissingColumns: bool,
            ?rechunk: bool,
            ?cache: bool,
            ?rowIndexName: string,
            ?rowIndexOffset: uint32,
            ?includePathColumn: string,
            ?schema: PolarsSchema,
            ?hivePartitioning:bool,
            ?hivePartitionSchema: PolarsSchema,
            ?tryParseHiveDates: bool
        ) : DataFrame =
            
            use ms = new System.IO.MemoryStream()
            stream.CopyTo ms
            let bytes = ms.ToArray()

            DataFrame.ReadParquet(
                buffer = bytes,
                ?columns = columns,
                ?nRows = nRows,
                ?parallelStrategy = parallelStrategy,
                ?lowMemory = lowMemory,
                ?useStatistics = useStatistics,
                ?allowMissingColumns = allowMissingColumns,
                ?rechunk = rechunk,
                ?cache = cache,
                ?rowIndexName = rowIndexName,
                ?rowIndexOffset = rowIndexOffset,
                ?includePathColumn = includePathColumn,
                ?schema = schema,
                ?hivePartitioning = hivePartitioning,
                ?hivePartitionSchema = hivePartitionSchema,
                ?tryParseHiveDates = tryParseHiveDates
            ) 

        /// <summary> Asynchronously read a Parquet file. </summary>
        static member ReadParquetAsync (        
                path: string,
                ?columns: seq<string>,
                ?nRows: uint64,
                ?parallelStrategy: ParallelStrategy,
                ?lowMemory: bool,
                ?useStatistics: bool,
                ?glob: bool,
                ?allowMissingColumns: bool,
                ?rechunk: bool,
                ?cache: bool,
                ?rowIndexName: string,
                ?rowIndexOffset: uint32,
                ?includePathColumn: string,
                ?schema: PolarsSchema,
                ?hivePartitioning:bool,
                ?hivePartitionSchema: PolarsSchema,
                ?tryParseHiveDates: bool,
                ?cloudOptions: CloudOptions) = 
            task {
                return DataFrame.ReadParquet(
                    path,
                    ?columns = columns,
                    ?nRows = nRows,
                    ?parallelStrategy = parallelStrategy,
                    ?lowMemory = lowMemory,
                    ?useStatistics = useStatistics,
                    ?glob = glob,
                    ?allowMissingColumns = allowMissingColumns,
                    ?rechunk = rechunk,
                    ?cache = cache,
                    ?rowIndexName = rowIndexName,
                    ?rowIndexOffset = rowIndexOffset,
                    ?includePathColumn = includePathColumn,
                    ?schema = schema,
                    ?hivePartitioning = hivePartitioning,
                    ?hivePartitionSchema = hivePartitionSchema,
                    ?tryParseHiveDates = tryParseHiveDates,
                    ?cloudOptions = cloudOptions
                )
            }
